game-core/src/main/java/games/strategy/triplea/delegate/battle/end/NoMoreUnits.java

```diff
@@ -0,0 +1,77 @@
+package games.strategy.triplea.delegate.battle.end;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.Matches;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Predicate;
+import lombok.Builder;
+import lombok.NonNull;
+import org.triplea.java.collections.CollectionUtils;
+
+@Builder
+public class NoMoreUnits {
+
+  public enum Winner {
+    ATTACKER,
+    DEFENDER,
+    DEFENDER_TURN_1,
+    UNDETERMINED
+  }
+
+  private @NonNull final GameData gameData;
+  private @NonNull final GamePlayer attacker;
+  private @NonNull final Collection<Unit> attackingUnits;
+  private @NonNull final Collection<Unit> defendingUnits;
+  private @NonNull final Territory battleSite;
+  private final int round;
+  private @NonNull final Runnable checkUndefendedTransports;
+  private @NonNull final Runnable checkForUnitsThatCanRollLeft;
+
+  public Winner check() {
+    final Predicate<Unit> notInfrastructure = Matches.unitIsNotInfrastructure();
+    if (CollectionUtils.getMatches(attackingUnits, notInfrastructure).isEmpty()) {
+      return noAttackingUnits();
+    } else if (CollectionUtils.getMatches(defendingUnits, notInfrastructure).isEmpty()) {
+      return noDefendingUnits();
+    }
+    return Winner.UNDETERMINED;
+  }
+
+  private Winner noAttackingUnits() {
+    final Winner result;
+    if (!Properties.getTransportCasualtiesRestricted(gameData)) {
+      result = Winner.DEFENDER;
+    } else {
+      // Get all allied transports in the territory
+      final Predicate<Unit> matchAllied =
+          Matches.unitIsTransport()
+              .and(Matches.unitIsNotCombatTransport())
+              .and(Matches.isUnitAllied(attacker, gameData));
+      final List<Unit> alliedTransports =
+          CollectionUtils.getMatches(battleSite.getUnits(), matchAllied);
+      // If no transports, just end the battle
+      if (alliedTransports.isEmpty()) {
+        result = Winner.DEFENDER;
+      } else if (round <= 1) {
+        result = Winner.DEFENDER_TURN_1;
+      } else {
+        result = Winner.DEFENDER;
```

Comment:
Is this correct to return `Winner.DEFENDER` here?

If so, then the matches for transport could be skipped and you'd only need to check if we are in round 1. It seems like this should be `UNDETERMINED`.

Reply:
That is what the original code had.  Are you sure that the transport check is no longer needed?  I don't see why.

Reply:
oh, now I see what you mean.  You're right.  It does look like it might be a bug but the original code is 
```
                if (alliedTransports.isEmpty()) {
                  endBattle(bridge);
                  defenderWins(bridge);
                } else if (round <= 1) {
                  attackingUnits =
                      CollectionUtils.getMatches(
                          battleSite.getUnits(), Matches.unitIsOwnedBy(attacker));
                } else {
                  endBattle(bridge);
                  defenderWins(bridge);
                }
```
I went through the history and that `else` statement has been around for a long time.  It was added in 2009 at https://github.com/triplea-game/triplea/commit/aa693839b8115e71a5b50ade06783a9657e10cdf.

I can simplify the if statements.

Reply:
On second look, the equivalent if is something like:
```
!alliedTransports.isEmpty() && round <=1
```

Perhaps we need the transport checking still. Though it's still pretty curious all in all.. I can't say I fully grok why transports are being checked.

Reply:
I've gone through the git history and I think the alliedTransport check is a bug.  The code used to also do the undefended transport check and when that was refactored, this alliedTransport part forgotten.

I came up with a situation where this would happen in a game and played it out and got a weird state in the game.  The situation is where your ally has your land units in it and is next to a country that has a factory.  On the enemies turn, a ship is produced in the same sea zone.  When the production is after combat, the two ships (enemy ship and allied transport) stay around.  On your turn, you move a ship into the zone and initiate a combat.  If your ship dies in the first round, you get an option to retreat because this code added your transported land units to the attacking units and they aren't dead.  If you retreat, your transported land units are moved by themselves.  I believe that is a bug.

I think the alliedTransport check should be removed.

Reply:
Oddly that is good news, fixes a problem and is much simpler. It might be worthwhile to hold off on that specific update and follow up on it in a new a dedicated PR to that change. It is up to you (keeping PRs atomic, focused is generally a very good thing. Makes for easier review and if rolled back then not as much is caught up in the rollback. Generally changes happen faster with many small PRs compared to fewer large ones).

Reply:
I made the change.  The battletree calculator will be simpler with this fix.